File: injection.py
import re
from dataclasses import dataclass
from enum import IntEnum
from typing import List


class RiskLevel(IntEnum):
    LOW = 0
    MEDIUM = 1
    HIGH = 2
    CRITICAL = 3


@dataclass
class Finding:
    pattern: str
    message: str
    risk: RiskLevel
    match_snippet: str = ""


# Heuristic rules (education / demo) — not a substitute for parameterization
_PATTERNS: list[tuple[str, str, RiskLevel, str | None]] = [
    (r"(?i)\bOR\s+['\"]?\d*['\"]?=\s*['\"]?\d*['\"]?", "Classic tautology in boolean context", RiskLevel.CRITICAL, None),
    (r"(?i)\bOR\s+['\"]1['\"]\s*=\s*['\"]1['\"]", "Tautology: OR '1'='1'", RiskLevel.CRITICAL, None),
    (r"(?i)\bUNION\s+SELECT\b", "UNION-based injection (UNION SELECT)", RiskLevel.CRITICAL, None),
    (r"(?i);[\s\n]*(DROP|TRUNCATE|ALTER|CREATE)\b", "Stacked command after semicolon", RiskLevel.CRITICAL, None),
    (r"(?i)EXEC(\s+|\()", "Execute / exec-style payload", RiskLevel.CRITICAL, None),
    (r"(?i)\bxp_\w+", "Extended stored procedure (xp_*)", RiskLevel.HIGH, None),
    (r"(?i)\bSLEEP\s*\(|\bBENCHMARK\s*\(|\bWAITFOR\s+DELAY\b", "Time-based blind SQL patterns", RiskLevel.HIGH, None),
    (r"(?i)INTO\s+OUTFILE|LOAD_FILE\s*\(", "File read / write patterns", RiskLevel.HIGH, None),
    (r"(?i)/\*.*?\*/", "Block comment (often used to bypass filters)", RiskLevel.MEDIUM, None),
    (r"(?i)--\s*[^
]+|\#[^
]*", "Line comment in expression (often to truncate server-side query)", RiskLevel.MEDIUM, None),
    (r"(?i)\bINFORMATION_SCHEMA\b|\bSYS\.\b|\bpg_catalog\b", "Metadata catalog access", RiskLevel.MEDIUM, None),
    (r"(?i)\bCHAR\s*\(\s*\d", "CHAR() concatenation (obfuscation)", RiskLevel.MEDIUM, None),
    (r"(?i)CONCAT\s*\(|0x[0-9a-fA-F]+", "String concatenation / hex literals", RiskLevel.LOW, None),
]


def analyze_injection(text: str) -> dict:
    """
    Scan raw SQL for common injection-style patterns.
    
    This function analyzes SQL queries for potential injection vulnerabilities
    using regex-based pattern matching.
    
    Args:
        text (str): The SQL query string to analyze.
    
    Returns:
        dict: A dictionary containing:
            - "level" (str): Risk level as string (low, medium, high, critical)
            - "level_num" (int): Risk level as integer
            - "findings" (list): List of detected patterns with details
    """
    if not text or not str(text).strip():
        return {
            "level": "low",
            "level_num": int(RiskLevel.LOW),
            "findings": [],
        }
    s = str(text)
    findings: List[Finding] = []
    seen: set[str] = set()

    for regex, message, risk, _ in _PATTERNS:
        for m in re.finditer(regex, s, re.DOTALL):
            key = f"{m.start()}:{m.end()}:{message}"
            if key in seen:
                continue
            seen.add(key)
            snippet = m.group(0)
            if len(snippet) > 120:
                snippet = snippet[:117] + "..."
            findings.append(
                Finding(
                    pattern=regex,
                    message=message,
                    risk=risk,
                    match_snippet=snippet,
                )
            )

    max_risk = RiskLevel.LOW
    if findings:
        max_risk = max(f.risk for f in findings)
    level_names = {RiskLevel.LOW: "low", RiskLevel.MEDIUM: "medium", RiskLevel.HIGH: "high", RiskLevel.CRITICAL: "critical"}
    return {
        "level": level_names[max_risk],
        "level_num": int(max_risk),
        "findings": [
            {
                "pattern": f.pattern,
                "message": f.message,
                "risk": level_names[f.risk],
                "snippet": f.match_snippet,
            }
            for f in sorted(findings, key=lambda x: -int(x.risk))
        ],
    }

File: injection.c
#include "injection.h"
#include "driver.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    const char *needle;
    const char *message;
    const char *risk;
    int level_num;
} Pattern;

static const Pattern PATTERNS[] = {
    {"or '1'='1'", "Tautology: OR '1'='1'", "critical", 3},
    {"union select", "UNION-based injection (UNION SELECT)", "critical", 3},
    {";drop", "Stacked command after semicolon", "critical", 3},
    {";truncate", "Stacked command after semicolon", "critical", 3},
    {";alter", "Stacked command after semicolon", "critical", 3},
    {";create", "Stacked command after semicolon", "critical", 3},
    {"exec(", "Execute / exec-style payload", "critical", 3},
    {" xp_", "Extended stored procedure (xp_*)", "high", 2},
    {"sleep(", "Time-based blind SQL patterns", "high", 2},
    {"benchmark(", "Time-based blind SQL patterns", "high", 2},
    {"waitfor delay", "Time-based blind SQL patterns", "high", 2},
    {"into outfile", "File read / write patterns", "high", 2},
    {"load_file(", "File read / write patterns", "high", 2},
    {"/*", "Block comment (often used to bypass filters)", "medium", 1},
    {"--", "Line comment in expression", "medium", 1},
    {"#", "Line comment in expression", "medium", 1},
    {"information_schema", "Metadata catalog access", "medium", 1},
    {"pg_catalog", "Metadata catalog access", "medium", 1},
    {"char(", "CHAR() concatenation (obfuscation)", "medium", 1},
    {"concat(", "String concatenation / hex literals", "low", 0},
    {"0x", "String concatenation / hex literals", "low", 0},
};

static const int PATTERN_COUNT = (int)(sizeof(PATTERNS) / sizeof(PATTERNS[0]));

static char *to_lower_copy(const char *s)
{
    size_t i;
    size_t n = strlen(s);
    char *out = (char *)malloc(n + 1);
    if (!out) {
        return strdup("");
    }
    for (i = 0; i < n; i++) {
        char c = s[i];
        if (c >= 'A' && c <= 'Z') {
            out[i] = (char)(c - 'A' + 'a');
        } else {
            out[i] = c;
        }
    }
    out[n] = '\0';
    return out;
}

static char *json_escape_snippet(const char *s, size_t pos)
{
    size_t max_len = 120;
    size_t remain = strlen(s + pos);
    size_t len = remain > max_len ? max_len : remain;
    char temp[128];
    snprintf(temp, sizeof(temp), "%.*s%s", (int)len, s + pos, remain > max_len ? "..." : "");
    return dj_json_string(temp);
}

char *analyze_injection_json(const char *sql)
{
    int i;
    int highest = 0;
    int first = 1;
    char *lower;
    char *findings = strdup("[");
    if (!sql || !sql[0]) {
        return strdup("{\"level\":\"low\",\"level_num\":0,\"findings\":[]}");
    }
    lower = to_lower_copy(sql);

    for (i = 0; i < PATTERN_COUNT; i++) {
        char *hit = strstr(lower, PATTERNS[i].needle);
        if (hit) {
            size_t idx = (size_t)(hit - lower);
            char *snippet = json_escape_snippet(sql, idx);
            char *pattern_json = dj_json_string(PATTERNS[i].needle);
            char *message_json = dj_json_string(PATTERNS[i].message);
            char *entry = dj_aprintf(
                "%s{\"pattern\":%s,\"message\":%s,\"risk\":\"%s\",\"snippet\":%s}",
                first ? "" : ",",
                pattern_json,
                message_json,
                PATTERNS[i].risk,
                snippet);
            char *merged = dj_aprintf("%s%s", findings, entry);
            free(findings);
            findings = merged;
            free(entry);
            free(snippet);
            free(pattern_json);
            free(message_json);
            first = 0;
            if (PATTERNS[i].level_num > highest) {
                highest = PATTERNS[i].level_num;
            }
        }
    }

    {
        const char *level = "low";
        char *final_json;
        char *closed = dj_aprintf("%s]", findings);
        free(findings);
        if (highest == 1) {
            level = "medium";
        } else if (highest == 2) {
            level = "high";
        } else if (highest == 3) {
            level = "critical";
        }
        final_json = dj_aprintf("{\"level\":\"%s\",\"level_num\":%d,\"findings\":%s}", level, highest, closed);
        free(closed);
        free(lower);
        return final_json;
    }
}
